addons/mail/static/src/messaging/component/composer_text_input/composer_text_input.js

```diff
@@ -143,10 +248,83 @@ class ComposerTextInput extends Component {
         this._textareaRef.el.style.height = (this._textareaRef.el.scrollHeight) + "px";
     }
 
+    /**
+     * Validates a keyword in order to trigger mentions search.
+     *
+     * @private
+     * @param delimiter
+     * @param beginningOnly
+     * @return {*}
+     */
+    _validateKeyword(delimiter, beginningOnly) {
+        const leftString = this.getContent().substring(0, this._getSelectionStart());
+
+        // use position before delimiter because there should be whitespaces
+        // or line feed/carriage return before the delimiter
+        const beforeDelimiterPosition = leftString.lastIndexOf(delimiter) - 1;
+        if (beginningOnly && beforeDelimiterPosition > 0) {
+            return false;
+        }
+        let searchStr = this.getContent().substring(beforeDelimiterPosition, this._getSelectionStart());
+        // regex string start with delimiter or whitespace then delimiter
+        const pattern = "^"+delimiter+"|^\\s"+delimiter;
+        const regexStart = new RegExp(pattern, 'g');
+        // trim any left whitespaces or the left line feed/ carriage return
+        // at the beginning of the string
+        searchStr = searchStr.replace(/^\s\s*|^[\n\r]/g, '');
+        if (regexStart.test(searchStr) && searchStr.length) {
+            searchStr = searchStr.replace(pattern, '');
+            return searchStr.indexOf(' ') < 0 && !/[\r\n]/.test(searchStr)
+                ? searchStr.replace(delimiter, '')
+                : false;
+        }
+        return false;
+    }
+
+    /**
+     * Detects if mentioned partners are still in the composer text input
+     * and removes them if not.
+     *
+     * @private
+     */
+    _validatePartnerIds() {
+        const inputMentions = this.getContent().match(new RegExp("@"+'[^ ]+(?= |&nbsp;|$)', 'g'));
+        if (inputMentions) {
+            for (const id in this._mentions) {
+                const mention = this._mentions[id];
+                let inputMention = inputMentions.find(item => {
+                    return item === ("@" + mention.name).replace(/ /g, '\u00a0');
+                });
+                if (!inputMention) {
+                    this._mentions.splice(this._mentions.indexOf(mention), 1);
+                }
+            }
+        }
+        else {
+            this._mentions = [];
+        }
+    }
+
     //--------------------------------------------------------------------------
     // Handlers
     //--------------------------------------------------------------------------
 
+    /**
+     * @private
+     * @param ev
+     */
+    _onClickMentionItem(ev) {
+        ev.preventDefault();
+        let target = ev.currentTarget.attributes['data-id'];
+        if (!target) {
+            target = ev.currentTarget.offsetParent.attributes['data-id'];
+        }
+        const index = target.value;
```

Comment:
instead of relying on `ev.currentTarget` data, it should use sub-components and `trigger()` (or directly invoke store changes, that is call public method on composer entity with side-effects, here make changes on mention suggestion of this composer)